### PR TITLE
drivers: usb: device: fix Rx FIFO min size

### PR DESCRIPTION
**Pull request by Desvauxm-st**
the FIFO Rx need to have a Minimum memory to works distributed the rest of the ram_size memory between the different TX FIFOs except the first which is
a control endtype with max data payload of 64 bytes

### CODE REVIEW

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -437,11 +441,17 @@ static int usb_dc_stm32_init(void)
...
-	HAL_PCDEx_SetRxFiFo(&usb_dc_stm32_state.pcd, FIFO_EP_WORDS);
+	HAL_PCDEx_SetRxFiFo(&usb_dc_stm32_state.pcd, FIFO_RX_MIN);
 	for (i = 0U; i < USB_NUM_BIDIR_ENDPOINTS; i++) {
-		HAL_PCDEx_SetTxFiFo(&usb_dc_stm32_state.pcd, i,
-				    FIFO_EP_WORDS);
-		k_sem_init(&usb_dc_stm32_state.in_ep_state[i].write_sem, 1, 1);
+		if (i == 0) {
+			/* first endpoint need only 64 byte for EP_TYPE_CTRL */
+			HAL_PCDEx_SetTxFiFo(&usb_dc_stm32_state.pcd, i,	64);
+			k_sem_init(&usb_dc_stm32_state.in_ep_state[i].write_sem, 1, 1);
```
Comment (erwango) [inline drivers/usb/device/usb_dc_stm32.c:449]: `k_sem_init()` can be put after the `if/else`

File: drivers/usb/device/usb_dc_stm32.c [hunk 2]
```diff
@@ -127,14 +127,18 @@ static const struct gpio_dt_spec ulpi_reset =
...
 
-/* 4-byte words FIFO */
-#define FIFO_WORDS (USB_RAM_SIZE / 4)
+/* FIFO Rx Minimum memory to works */
+#define FIFO_RX_MIN 160
 
-/* Allocate FIFO memory evenly between the FIFOs */
-#define FIFO_EP_WORDS (FIFO_WORDS / FIFO_NUM)
+/* 4-byte words TX FIFO */
+#define TX_FIFO_WORDS ((USB_RAM_SIZE - FIFO_RX_MIN) / 4)
```
Comment (erwango) [inline drivers/usb/device/usb_dc_stm32.c:137]: Shouldn't we also take into account the 64 bytes assigned to EP0 in this ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_stm32.c (hunk 1 above)
Comment (Desvauxm-st) [reply drivers/usb/device/usb_dc_stm32.c:447]: yes

File: drivers/usb/device/usb_dc_stm32.c (hunk 2 above)
Comment (Desvauxm-st) [reply drivers/usb/device/usb_dc_stm32.c]: yes sorry

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: subsys/usb/device/Kconfig [hunk 3]
```diff
@@ -28,6 +28,12 @@ config USB_DEVICE_PID
 	help
 	  USB device product ID. MUST be configured by vendor.
 
+config USB_FIFO_RX_MIN
+	int "USB FIFO RX min"
+	default 160
+	help
+	  FIFO Rx Minimum memory to works. MUST be configured by vendor.
+
```
[the remark this answers is not in the document]
Comment (erwango) [reply subsys/usb/device/Kconfig:36]: What about adding this as a STM32 Kconfig ? (This is application tuning which is better suited as Kconfig than device tree)

[the remark this answers is not in the document]
Comment (jfischer-no) [reply subsys/usb/device/Kconfig:36]: This is certainly not application tuning.

[the remark this answers is not in the document]
Comment (erwango) [reply subsys/usb/device/Kconfig:36]: Ok, I was of course suggesting a vendor driver specific symbol in `drivers/usb/device/Kconfig`.

[the remark this answers is not in the document]
Comment (jfischer-no) [reply subsys/usb/device/Kconfig:36]: Still, there are open questions and it does not make any sense for devices with dedicated FIFOs.

[the remark this answers is not in the document]
Comment (erwango) [reply subsys/usb/device/Kconfig:36]: > Still, there are open questions and it does not make any sense for devices with dedicated FIFOs.

I specifically mentioned STM32. In that case, is that acceptable ?

File: drivers/usb/device/usb_dc_stm32.c [hunk 4]
```diff
@@ -436,11 +437,17 @@ static int usb_dc_stm32_init(void)
...
+
 	/* TODO: make this dynamic (depending usage) */
-	HAL_PCDEx_SetRxFiFo(&usb_dc_stm32_state.pcd, FIFO_EP_WORDS);
+	HAL_PCDEx_SetRxFiFo(&usb_dc_stm32_state.pcd, CONFIG_USB_FIFO_RX_MIN);
 	for (i = 0U; i < USB_NUM_BIDIR_ENDPOINTS; i++) {
-		HAL_PCDEx_SetTxFiFo(&usb_dc_stm32_state.pcd, i,
-				    FIFO_EP_WORDS);
+		if (i == 0) {
+			/* first endpoint need only 64 byte for EP_TYPE_CTRL */
+			HAL_PCDEx_SetTxFiFo(&usb_dc_stm32_state.pcd, i,	64);
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c:446]: You mentioned 64bytes, but the parameter seems to be in word count, not byte.

Comment (Desvauxm-st) [reply drivers/usb/device/usb_dc_stm32.c:446]: yes you right in fact this value is term of 32 bits : ( see documentation after: )

image

Comment (Desvauxm-st) [reply drivers/usb/device/usb_dc_stm32.c]: I push a new modification

File: subsys/usb/device/Kconfig (hunk 3 above)
[the remark this answers is not in the document]
Comment (tmon-nordic) [reply subsys/usb/device/Kconfig:36]: This is not really something that should be Kconfig-able. If it need to be possible for user to modify, then make it a devicetree property.

**Comment by loicpoulain** [comment]
Looks good, but can you address the @tmon-nordic comment about devicetree property for rx-fifo size if it makes sense? In that case, udc_stm32 should use it as well.

File: subsys/usb/device/Kconfig (hunk 3 above)
[the remark this answers is not in the document]
Comment (Desvauxm-st) [reply subsys/usb/device/Kconfig]: ok

**Review by loicpoulain — APPROVED** [review]

**Review by tmon-nordic — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]
LGTM, At least it does not make it worse, I am still curious what the benefit is if dynamic FIFO sizing or shared FIFO is not enabled for certain hardware configuration.

**Comment by tmon-nordic** [comment]
@erwango This PR shows that you have requested changes. If this is not intended then either approve or dismiss the old review.

**Review by erwango — APPROVED** [review]

**Pull request merged**